File: FrondEnd/src/app/core/servicios/progra.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { AuthServiceService } from './auth.service.service';
import { Progra, PageResponse } from '../models/auth-response.model';

@Injectable({ providedIn: 'root' })
export class PrograService {
  private apiURL = `${environment.apiURL}/api/progra`;
  private http = inject(HttpClient);
  private authService = inject(AuthServiceService);

  private getAuthHeaders(): HttpHeaders {
    return new HttpHeaders({ Authorization: `Bearer ${this.authService.getToken()}` });
  }

  listarPaginado(filtro: string, estado: string, codn: number, page: number, size: number): Observable<PageResponse<Progra>> {
    const pageZeroBased = page - 1;
    let params = new HttpParams()
      .set('filtro', filtro).set('estado', estado)
      .set('page', pageZeroBased.toString()).set('size', size.toString());
    if (codn > 0) params = params.set('codn', codn.toString());

    return this.http.get<PageResponse<Progra>>(this.apiURL, { headers: this.getAuthHeaders(), params });
  }

  crear(codmat: string, codpar: number, codp: number, gestion: number, login: string): Observable<Progra> {
    let params = new HttpParams()
      .set('codmat', codmat).set('codpar', codpar.toString())
      .set('codp', codp.toString()).set('gestion', gestion.toString()).set('login', login);
    return this.http.post<Progra>(this.apiURL, null, { headers: this.getAuthHeaders(), params });
  }

  modificar(idViejo: Progra['id'], codmatNew: string, codparNew: number, codpNew: number, login: string): Observable<Progra> {
    let params = new HttpParams()
      .set('oldCodmat', idViejo.codmat).set('oldCodpar', idViejo.codpar.toString())
      .set('oldCodp', idViejo.codp.toString()).set('oldGestion', idViejo.gestion.toString())
      .set('newCodmat', codmatNew).set('newCodpar', codparNew.toString())
      .set('newCodp', codpNew.toString()).set('login', login);
    return this.http.put<Progra>(this.apiURL, null, { headers: this.getAuthHeaders(), params });
  }

  eliminar(id: Progra['id']): Observable<void> {
    let params = new HttpParams()
      .set('codmat', id.codmat).set('codpar', id.codpar.toString())
      .set('codp', id.codp.toString()).set('gestion', id.gestion.toString());
    return this.http.delete<void>(this.apiURL, { headers: this.getAuthHeaders(), params });
  }

  habilitar(id: Progra['id']): Observable<void> {
    let params = new HttpParams()
      .set('codmat', id.codmat).set('codpar', id.codpar.toString())
      .set('codp', id.codp.toString()).set('gestion', id.gestion.toString());
    return this.http.put<void>(`${this.apiURL}/habilitar`, null, { headers: this.getAuthHeaders(), params });
  }
}

File: src/main/java/com/Proyecto/backEnd/service/PrograService.java
package com.Proyecto.backEnd.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import com.Proyecto.backEnd.model.*;
import com.Proyecto.backEnd.repository.*;
import jakarta.persistence.criteria.*;
import jakarta.transaction.Transactional;

@Service
public class PrograService {

    @Autowired
    private PrograRepo prograRepo;
    @Autowired
    private UsuariosRepo usuariosRepo;
    @Autowired
    private MateriasRepo materiasRepo;
    @Autowired
    private ParalelosRepo paralelosRepo;
    @Autowired
    private PersonalRepo personalRepo;

    /**
     * B-17. Listar, filtrar y paginar Inscripciones
     */
    public Page<PrograModel> listarPaginado(String filtro, String estado, Integer codn, Pageable pageable) {

        Specification<PrograModel> spec = (root, query, cb) -> {
            root.fetch("materia").fetch("nivel");
            root.fetch("alumno");
            root.fetch("paralelo");
            query.distinct(true);

            Predicate p = cb.conjunction();

            if (estado.equals("ACTIVOS")) {
                p = cb.and(p, cb.equal(root.get("estado"), 1));
            } else if (estado.equals("BAJAS")) {
                p = cb.and(p, cb.equal(root.get("estado"), 0));
            }

            if (codn != null && codn > 0) {
                Join<PrograModel, MateriasModel> materiaJoin = root.join("materia");
                Join<MateriasModel, NivelesModel> nivelJoin = materiaJoin.join("nivel");
                p = cb.and(p, cb.equal(nivelJoin.get("codn"), codn));
            }

            if (filtro != null && !filtro.isEmpty()) {
                String filtroLower = "%" + filtro.toLowerCase() + "%";

                Join<PrograModel, PersonalModel> alumnoJoin = root.join("alumno");
                Join<PrograModel, MateriasModel> materiaJoin = root.join("materia", JoinType.LEFT);

                Predicate filtroAlumno = cb.or(
                        cb.like(cb.lower(alumnoJoin.get("nombre")), filtroLower),
                        cb.like(cb.lower(alumnoJoin.get("ap")), filtroLower));
                Predicate filtroMateria = cb.like(cb.lower(materiaJoin.get("nombre")), filtroLower);

                p = cb.and(p, cb.or(filtroAlumno, filtroMateria));
            }
            return p;
        };
        return prograRepo.findAll(spec, pageable);
    }

    /**
     * B-17.1. Crear Inscripción
     */
    public PrograModel crearInscripcion(String codmat, int codpar, int codp, int gestion, String login) {

        PrograId id = new PrograId(codpar, codp, codmat, gestion);
        if (prograRepo.existsById(id)) {
            return null; // Ya existe
        }

        MateriasModel mat = materiasRepo.findById(codmat)
                .orElseThrow(() -> new RuntimeException("Materia no encontrada"));
        ParalelosModel par = paralelosRepo.findById(codpar)
                .orElseThrow(() -> new RuntimeException("Paralelo no encontrado"));
        PersonalModel alum = personalRepo.findById(codp)
                .orElseThrow(() -> new RuntimeException("Alumno no encontrado"));
        UsuariosModel usu = usuariosRepo.findById(login)
                .orElseThrow(() -> new RuntimeException("Usuario no encontrado"));

        PrograModel nueva = new PrograModel();
        nueva.setId(id);
        nueva.setMateria(mat);
        nueva.setParalelo(par);
        nueva.setAlumno(alum);
        nueva.setUsuario(usu);
        nueva.setEstado(1);

        return prograRepo.save(nueva);
    }

    /**
     * B-17.2. Modificar Inscripción (Delete + Create)
     */
    @Transactional
    public PrograModel modificarInscripcion(
            String oldCodmat, int oldCodpar, int oldCodp, int oldGestion,
            String newCodmat, int newCodpar, int newCodp, String login) {

        PrograId idViejo = new PrograId(oldCodpar, oldCodp, oldCodmat, oldGestion);
        if (!prograRepo.existsById(idViejo)) {
            throw new RuntimeException("No se encontró la inscripción original.");
        }
        prograRepo.deleteById(idViejo);

        PrograModel nueva = this.crearInscripcion(newCodmat, newCodpar, newCodp, oldGestion, login);
        if (nueva == null) {
            throw new RuntimeException("La nueva inscripción ya existe.");
        }
        return nueva;
    }

    public void eliminarLogico(PrograId id) {
        PrograModel p = prograRepo.findById(id).orElseThrow(() -> new RuntimeException("Inscripción no encontrada"));
        p.setEstado(0);
        prograRepo.save(p);
    }

    public void habilitar(PrograId id) {
        PrograModel p = prograRepo.findById(id).orElseThrow(() -> new RuntimeException("Inscripción no encontrada"));
        p.setEstado(1);
        prograRepo.save(p);
    }
}